nengo/cache.py

```diff
@@ -272,45 +412,45 @@ class DecoderCache(object):
     """
 
     _CACHE_EXT = '.nco'
-    _INDEX = 'index'
-    _LEGACY = 'legacy.txt'
-    _LEGACY_VERSION = 1
-    _PICKLE_PROTOCOL = pickle.HIGHEST_PROTOCOL
 
     def __init__(self, readonly=False, cache_dir=None):
         self.readonly = readonly
         if cache_dir is None:
             cache_dir = self.get_default_dir()
         self.cache_dir = cache_dir
-        safe_makedirs(self.cache_dir)
+        if readonly:
+            self._index = CacheIndex(cache_dir)
+        else:
+            safe_makedirs(self.cache_dir)
+            self._index = WriteableCacheIndex(cache_dir)
         self._fragment_size = get_fragment_size(self.cache_dir)
-        self._index = None
         self._fd = None
+        self._in_context = False
 
     def __enter__(self):
         try:
-            self._remove_legacy_files()
-            index_path = os.path.join(self.cache_dir, self._INDEX)
-            self._index = CacheIndex(index_path)
             try:
                 self._index.__enter__()
-            except (EOFError, IOError, OSError):
-                # Index corrupted, clear cache because we can't recover
-                # information necessary to access cache data.
-                self.invalidate()
-                os.remove(index_path)
+            except TimeoutError:
+                self.readonly = True
+                self._index = CacheIndex(self.cache_dir)
                 self._index.__enter__()
-        except TimeoutError:
-            warnings.warn(
-                "Decoder cache could not acquire lock and was deactivated.")
+                warnings.warn(
+                    "Decoder cache could not acquire lock and was to "
+                    "readonly.")
+        except:
```

Comment:
I don't think so. No matter what exception occurs, this function should fail gracefully. (Otherwise some random cache initialization failure would kill all of your simulation.)